source/common/init/target_impl.cc

```diff
@@ -25,8 +25,13 @@ bool TargetHandleImpl::initialize(const Watcher& watcher) const {
 TargetImpl::TargetImpl(absl::string_view name, InitializeFn fn)
     : name_(fmt::format("target {}", name)),
       fn_(std::make_shared<InternalInitalizeFn>([this, fn](WatcherHandlePtr watcher_handle) {
-        watcher_handle_ = std::move(watcher_handle);
-        fn();
+        if (is_ready_) {
+          fn();
```

Comment:
I still don't really understand why the call to `fn()` is needed in the case where the target is already initialized.

Reply:
IMHO it's not a must-have. I leave it in this way because that is what user will see.
Suppose we have the code and happens to be executed in below order:
```
Context A                         Context B                         Context C
ManagerImpl m;
                                 TargetImpl t("foo", fn);
                                 m.add(t);
                                 ...                  
X:                                                                      t.ready();
Y: m.initialize(w);
```

Prior to this PR: 
1. Either t.ready() happens before m.initialize(w), or the reverse, we will observe that fn will be called. 
2. w will be triggered only if t.ready() happens after m.initialize(w)

This goal of this PR is to change item 2.

Our conflict is whether 1 should be maintained. I don't have a strong preference but I feel there is no gain to change 1.

Reply:
That's a really good explanation.

My assumption has always been that `fn()` (or something called by it indirectly) calls `t.ready()`. Or, in other words, if `t.ready()` has been called, there's no need for `fn()` to be called again. You've said that in your case, `fn` is actually empty, and `ready` is called by some other code, and I still sort of think that's an abuse of the init manager, but I guess we can ignore that for now :)

Here's my thinking: we can't assume that `fn` is idempotent, and I think we should assume that if `is_ready_` is true, that means initialization is done. Like, I think of it as really being `is_initialized_`. So, to your point, it would be a change in behavior to not call `fn`, but I think it's most consistent with the design intent of the init manager (that is, I think it's really fixing a latent bug).

Reply:
> but I guess we can ignore that for now :)

I haven't been fully tracking this conversation, but I don't think we should ignore this for now. Can we step back and make sure we are actually solving the right problem? However we think it's best to do that is fine with me.

Reply:
> we can't assume that fn is idempotent

I used to consider fn() 
1. a signal which is harmless to call more than once. 
2. may have side affect other than transit state from not-ready to ready. 
It leads to my code that InitManager should call it at least once.

I start to understand your concern:
if `is_ready` is true, it doesn't make sense that the init system calls fn().

Will update following your opinion